### PR TITLE
Overhauls discord verification system

### PR DESCRIPTION
**Pull request by optimumtact**
This completely replaces the previous verification system, for one that
will interoperate with a discord redbot instance that uses the cogs
located at you github.com/optimumtact/orangescogs

This cuts out several steps in the system, but it also leaves alone the
existing notify system (which just uses a file list of discord ids) as a
record of who to notify

SQL changes required for the new database system

Version 5.10, 7 August 2020, by oranges

Changes how the discord verification process works.
Adds the discord_links table, and migrates discord id entries from
player table to the discord links table in a once off operation and then
removes the discord id on the player table

The user connects to any tg server, and uses the "Verify Discord
Account" verb, this generates a six word one time use token, with a 4
hour time validity period (defined as 4 hours from the timestamp value)
in the discord links table.

This one time token, and the ckey of the user are stored in
discord_links

At this point the entire DM side is done, this is all it does

The rest is done by the cogs.

## Why It's Good For The Game
Provides a way to reduce ban evasion and bad actors on discord, as they must have a certain amount
of living time to be let into the discord

### CODE REVIEW

File: code/controllers/subsystem/discord.dm [hunk 1]
```diff
@@ -36,15 +39,22 @@ SUBSYSTEM_DEF(discord)
...
-	/// List that holds accounts to link, used in conjunction with TGS
-	var/list/account_link_cache = list()
-	/// list of people who tried to reverify, so they can only do it once per round as a shitty slowdown
-	var/list/reverify_cache = list()
+
+	/// People who have tried to verify this round already
+	var/list/reverify_cache
+
+	/// Common words list, used to generate one time tokens
+	var/static/list/common_words
```
Comment (optimumtact) [inline code/controllers/subsystem/discord.dm:47]: just realised this probably doesn't work, since it's static? and initiated later?

**Comment by ATH1909** [comment]
I assume that this is directed at mischief? and his ilk?

**Comment by optimumtact** [comment]
No, it's more that we had this, but it was too complex

**Comment by ATH1909** [comment]
> No, it's more that we had this, but it was too complex

Ah, fair enough.

File: code/controllers/subsystem/discord.dm (hunk 1 above)
Comment (Jared-Fogle) [reply code/controllers/subsystem/discord.dm:47]: Why static anyway if subsystems are only initialized once?

Comment (optimumtact) [reply code/controllers/subsystem/discord.dm:47]: I dunno, paranoia, faint hope it's more effective?

Comment (Rohesie) [reply code/controllers/subsystem/discord.dm:47]: Just make it not-static and it should work as intended.

File: code/controllers/subsystem/discord.dm [hunk 2]
```diff
@@ -81,70 +91,184 @@ SUBSYSTEM_DEF(discord)
...
+  *
+  * Arguments:
+  * * ckey_for a string representing the ckey this token is for
+  *
+  * Returns a string representing the one time token
+  */
+/datum/controller/subsystem/discord/proc/generate_one_time_token(ckey_for)
+
+	var not_unique = TRUE
+	var one_time_token = ""
```
Comment (Rohesie) [inline code/controllers/subsystem/discord.dm:155]:
```suggestion
	var/not_unique = TRUE
	var/one_time_token = ""
```

**Review by AffectedArc07 — APPROVED** [review]
Seems fine for now, just make sure you dont break existing notify functionality.

File: code/controllers/subsystem/discord.dm [hunk 3]
```diff
@@ -81,70 +91,184 @@ SUBSYSTEM_DEF(discord)
...
+  * Remember, multiple entries can exist
+  *
+  * Arguments:
+  * * discordid The users discord id (string)
+  * * timebound should we search only in the last 4 hours
+  *
+  * Returns a [/datum/discord_link_record]
+  */
+/datum/controller/subsystem/discord/proc/find_discord_link_by_discord_id(discord_id, timebound = FALSE)
+	var timeboundsql = ""
```
Comment (AffectedArc07) [inline code/controllers/subsystem/discord.dm:256]:
```suggestion
	var/timeboundsql = ""
```

File: code/controllers/subsystem/discord.dm [hunk 4]
```diff
@@ -81,70 +91,184 @@ SUBSYSTEM_DEF(discord)
...
+	var not_unique = TRUE
+	var one_time_token = ""
+	// While there's a collision in the token, generate a new one (should rarely happen)
+	while(not_unique)
+		one_time_token = "[pick(common_words)]-[pick(common_words)]-[pick(common_words)]-[pick(common_words)]-[pick(common_words)]-[pick(common_words)]"
+		not_unique = find_discord_link_by_token(one_time_token, timebound = TRUE)
+
+	// Insert into the table, null in the discord id, id and timestamp fields so the db fills them out where needed
+	var/datum/db_query/query_insert_link_record = SSdbcore.NewQuery(
+		"INSERT INTO [format_table_name("discord_link")] VALUES(NULL, :ckey, NULL, NULL, :token)",
```
Comment (Jordie0608) [inline code/controllers/subsystem/discord.dm:163]: Don't use the insert shorthand
```suggestion
		"INSERT INTO [format_table_name("discord_link")] (ckey, token) VALUES(:ckey, :token)",
```

File: code/controllers/subsystem/discord.dm [hunk 5]
```diff
@@ -81,70 +91,184 @@ SUBSYSTEM_DEF(discord)
...
+  * * timebound should we search only in the last 4 hours
+  *
+  * Returns a [/datum/discord_link_record]
+  */
+/datum/controller/subsystem/discord/proc/find_discord_link_by_ckey(ckey, timebound = FALSE)
+	var timeboundsql = ""
+	if(timebound)
+		timeboundsql = "AND timestamp >= Now() - INTERVAL 4 HOUR"
+
+	var query = "SELECT CAST(discord_id AS CHAR(25)), ckey, MAX(timestamp), one_time_token FROM [format_table_name("discord_link")] WHERE ckey = :ckey [timeboundsql] GROUP BY ckey, discord_id, one_time_token",
```
Comment (Jordie0608) [inline code/controllers/subsystem/discord.dm:225]: If this is only looking for the first result, use `LIMIT 1`

File: code/controllers/subsystem/discord.dm [hunk 6]
```diff
@@ -81,70 +91,184 @@ SUBSYSTEM_DEF(discord)
...
+  * * timebound should we search only in the last 4 hours
+  *
+  * Returns a [/datum/discord_link_record]
+  */
+/datum/controller/subsystem/discord/proc/find_discord_link_by_discord_id(discord_id, timebound = FALSE)
+	var timeboundsql = ""
+	if(timebound)
+		timeboundsql = "AND timestamp >= Now() - INTERVAL 4 HOUR"
+
+	var query = "SELECT CAST(discord_id AS CHAR(25)), ckey, MAX(timestamp), one_time_token FROM [format_table_name("discord_link")] WHERE discord_id = :discord_id [timeboundsql] GROUP BY ckey, discord_id, one_time_token",
```
Comment (Jordie0608) [inline code/controllers/subsystem/discord.dm:260]: Same.

**Review by Jordie0608 — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/discord.dm (hunk 2 above)
Comment (Rohesie) [reply code/controllers/subsystem/discord.dm:155]: Though it works, we don't use this syntax on DM (the only instances we have are js parts  of it)
It's in the contributing guidelines as absolute pathing.

File: SQL/database_changelog.txt [hunk 7]
```diff
@@ -1,15 +1,39 @@
...
+
+BEGIN;
+
+DROP TABLE IF EXISTS `discord_links`;
+CREATE TABLE `discord_links` (
+	`id` int(11) NOT NULL AUTO_INCREMENT,
+	`ckey` VARCHAR(32) NOT NULL,
+	`discord_id` BIGINT(20) DEFAULT NULL,
+	`timestamp` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+	`one_time_token` TEXT NOT NULL,
```
[the remark this answers is not in the document]
Comment (optimumtact) [reply SQL/database_changelog.txt:27]: I didn't want to find the largest word in the 1000 common words and multiply it by 6

[the remark this answers is not in the document]
Comment (Jordie0608) [reply SQL/database_changelog.txt:27]: The longest words are `conversation` and `relationship` at 12 characters so with hyphens the max possible length is 77.

File: SQL/database_changelog.txt [hunk 8]
```diff
@@ -1,15 +1,39 @@
...
 
 -----------------------------------------------------
 
+Version 5.10, 7 August 2020, by oranges
+
+Changes how the discord verification process works.
+Adds the discord_links table, and migrates discord id entries from player table to the discord links table in a once off operation and then removes the discord id
+on the player table
+
+BEGIN;
```
[the remark this answers is not in the document]
Comment (optimumtact) [reply SQL/database_changelog.txt:19]: okay

File: SQL/database_changelog.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (optimumtact) [reply SQL/database_changelog.txt:27]: thanks I'll make it a varchar 100

**Comment by optimumtact** [comment]
Need to add a "valid" flag, so admins can disable verification status of a banned user if they are also being a pain in discord.

- [ ] add validity flag and check it in all related DB queries.

**Comment by AffectedArc07** [comment]
May be worth updating https://github.com/tgstation/tgstation/tree/master/tools/discordRoleScript to the new schema, but then again this is for when giving roles was DM side, not discord-side.

**Comment by optimumtact** [comment]
read for review

File: code/controllers/subsystem/discord.dm [hunk 9]
```diff
@@ -81,70 +91,186 @@ SUBSYSTEM_DEF(discord)
...
+  * Remember, multiple entries can exist, as they are only guaranteed to be unique for their validity period
+  *
+  * Arguments:
+  * * one_time_token the string of words representing the one time token
+  * * timebound A boolean flag, that specifies if it should only look for entries within the last 4 hours, off by default
+  *
+  * Returns a [/datum/discord_link_record]
+  */
+/datum/controller/subsystem/discord/proc/find_discord_link_by_token(one_time_token, timebound = FALSE)
+	var timeboundsql = ""
```
Comment (Rohesie) [inline code/controllers/subsystem/discord.dm:191]:
```suggestion
	var/timeboundsql = ""
```

File: code/controllers/subsystem/discord.dm [hunk 10]
```diff
@@ -81,70 +91,186 @@ SUBSYSTEM_DEF(discord)
...
+  * Remember, multiple entries can exist
+  *
+  * Arguments:
+  * * ckey the users ckey as a string
+  * * timebound should we search only in the last 4 hours
+  *
+  * Returns a [/datum/discord_link_record]
+  */
+/datum/controller/subsystem/discord/proc/find_discord_link_by_ckey(ckey, timebound = FALSE)
+	var timeboundsql = ""
```
Comment (Rohesie) [inline code/controllers/subsystem/discord.dm:223]:
```suggestion
	var/timeboundsql = ""
```

File: code/modules/discord/accountlink.dm [hunk 11]
```diff
@@ -1,93 +1,30 @@
...
 	// Safety checks
 	if(!CONFIG_GET(flag/sql_enabled))
 		to_chat(src, "<span class='warning'>This feature requires the SQL backend to be running.</span>")
 		return
 
-	// ss is still starting
-	if(!SSdiscord)
-		to_chat(src, "<span class='notice'>The server is still starting up. Please wait before attempting to link your account!</span>")
-		return
+	var message = ""
```
Comment (Rohesie) [inline code/modules/discord/accountlink.dm:12]:
```suggestion
	var/message = ""
```

File: code/modules/discord/accountlink.dm [hunk 12]
```diff
@@ -1,93 +1,30 @@
...
+	// Simple sanity check to prevent a user doing this too often
+	var cached_one_time_token = SSdiscord.reverify_cache[usr.ckey]
```
Comment (Rohesie) [inline code/modules/discord/accountlink.dm:14]:
```suggestion
	var/cached_one_time_token = SSdiscord.reverify_cache[usr.ckey]
```

File: code/modules/discord/accountlink.dm [hunk 13]
```diff
@@ -1,93 +1,30 @@
...
-		return
 
-	// check that this is not an IDIOT mistaking us for an attack vector
-	if(SSdiscord.reverify_cache[usr.ckey] == TRUE)
-		to_chat(src, "<span class='warning'>Thou can only do this once a round, if you're stuck seek help.</span>")
-		return
-	SSdiscord.reverify_cache[usr.ckey] = TRUE
+	else
+		// Will generate one if an expired one doesn't exist already, otherwise will grab existing token
+		var one_time_token = SSdiscord.get_or_generate_one_time_token_for_ckey(ckey)
```
Comment (Rohesie) [inline code/modules/discord/accountlink.dm:21]:
```suggestion
		var/one_time_token = SSdiscord.get_or_generate_one_time_token_for_ckey(ckey)
```

**Comment by optimumtact** [comment]
I boomed myself.

**Comment by optimumtact** [comment]
YOU can't DO THIS, I HAVE RIGHTS

**Pull request closed without merging**